# Assigment 3
---
### Sudoku as a constraint Satisfaction Problem
Sudoku is one of the most popular puzzle games of all time. The goal of Sudoku is to fill in a 9×9 grid with digits so that each column, row, and 3×3 section contain the numbers between 1 to 9. At the beginning of the game, the 9×9 grid will have some of the squares filled in. Your job is to use logic to fill in the missing digits and complete the grid. A move is incorrect if:

    •	Any row contains more than one of the same number from 1 to 9

    •	Any column contains more than one of the same number from 1 to 9

    •	Any 3×3 grid contains more than one of the same number from 1 to 9

<p align="center">
<img src="https://i.stack.imgur.com/vuMhM.gif" width="250" height="250" /></p>

A Sudoku text file should contain the information of the Sudoku puzzle in the following format.

    Sudoku 01
    003020600
    900305001
    001806400
    008102900
    700000008
    006708200
    002609500
    800203009
    005010300

### Task 1 : Parse the Sudoku data file (5 pts)

### Task 2 : Naïve Backtracking Algorithm (30 pts)
Implement a naïve backtracking algorithm. The selection of variables and assignment of values can be done either in order or randomly.

### Task 3 : Smart Backtracking Algorithm (40 pts)
Incorporate at least one strategy of minimum remaining values (MRV), least constraining value (LCV), and forward checking in your backtracking algorithm.

### Task 4 : Report and Analysis (25 pts)

The following website provides Sudoku puzzles in levels of easy, medium, hard, and evil. Analyze the performance of your Sudoku solver on these puzzles.
http://www.websodoku.com

In [2]:
#--- Importing Libraries
import numpy as np
import io
import math
from functools import reduce

In [87]:
#--- Paths for sudoku test files
# Test1     ------ easy
# Test2     ------ medium
# TestFinal ------ hard
# evil      ------ evil

data = "./testcases/sudoku-01.txt"
test2 = "./testcases/sudoku-test-2.txt"
test1 = "./testcases/sudoku-test-1.txt"
evil = "./testcases/sudoku-evil-1.txt"
testFinal = "./testcases/sudoku-final-test.txt"

def getPuzzle(filename):
    with open(filename, 'r') as f:
        lines = f.read().split('\n')
    content = list()
    for line in lines:
        if len(line) != 0:
            line = list(str(line))
            content.append(line)
    puzzle = np.array(content)
    puzzle = puzzle[1:][:]
    puzzle = puzzle.astype(int)
    print(puzzle)
    return puzzle

puzzle = getPuzzle(test2)

[[3 0 6 5 0 8 4 0 0]
 [5 2 0 0 0 0 0 0 0]
 [0 8 7 0 0 0 0 3 1]
 [0 0 3 0 1 0 0 8 0]
 [9 0 0 8 6 3 0 0 5]
 [0 5 0 0 9 0 6 0 0]
 [1 3 0 0 0 0 2 5 0]
 [0 0 0 0 0 0 0 7 4]
 [0 0 5 2 0 6 3 0 0]]


### Task 2 : Naïve Backtracking Algorithm (30 pts)
Implement a naïve backtracking algorithm. The selection of variables and assignment of values can be done either in order or randomly.

<p align="center">
<img src="https://codemyroad.files.wordpress.com/2014/04/output_hqxeh9.gif" width="250" height="250" /></p>



In [88]:
#--- sudoku solver
def startSolving(row, col, puzzle):
    #--- terminating condition when we reach end of the puzzle
    if col == len(puzzle[row]):               #--- check if it has reached end of the column, if yes increment the row
        col = 0
        row+=1    
        if row == len(puzzle):                  #--- check if it has reached end of the board if yes return True and puzzle
            return True
    #--- skip filled positions, go to next position
    if puzzle[row][col] != 0 :
        return startSolving(row,col+1,puzzle)   #--- recursion
    #--- for each row check if we can place the value in each position of the row
    for i in range(1,len(puzzle)+1):          #--- iterate from 1 - 9 and check if we can place the value
        #print(i)
        if canPlace(puzzle, row, col, i):       #--- checks all the constraints
            puzzle[row, col] = i;
            if startSolving(row, col+1, puzzle):  #--- once the constraint is passed and element is placed move on to next element
                return True
        puzzle[row, col] = 0;                 #--- if the constraint is not passed then  place the value zero
    return False

#--- check if the value satisfies all the constraints
def canPlace(puzzle, row, col, value):
    if checkRow(puzzle, row, value):
        if checkCol(puzzle, col, value):
              if checkGrid(puzzle, row, col, value):
                    return True                        #--- Return True as all the constrainst are satisfied
    return False                             #--- Return False If any constrainst fails

def checkRow(puzzle, row, value):
    #--- check for row 
    if value in puzzle[row,:]:
        return False                         #--- Return False as there is same element present in the row      
    return True

def checkCol(puzzle, col, value):
  #--- check for column 
  if value in puzzle[:,col]:
    return False                           #--- Return False as there is same element present in the column
  return True

def checkGrid(puzzle, row, col, value):
    #--- check for the 3x3 grid
    gridsize = int(math.sqrt(len(puzzle)))
    #--- break the puzzle into grids of gridsize
    vertical = row // gridsize
    horizontal = col//gridsize

    #--- get start of grid
    gridstartRow = gridsize*vertical
    gridstartCol = gridsize*horizontal

    #--- check the grid
    for i in range(0,gridsize):
        for j in range(0,gridsize):
            if value == puzzle[gridstartRow+i,gridstartCol+j]:
                return False                      #--- Return False as there is same element present in the column
    return True

def solveSudoku(puzzle):
    if startSolving(0,0,puzzle):            #--- start solving sudoku from initial position on the board
        print(puzzle)                         #--- print the result if the puzzle is solved
    else:
        print('Unable to solve the puzzle')  #--- else print the message that unable to solve the puzzle

In [89]:
%%time
solveSudoku(puzzle)

[[3 1 6 5 7 8 4 9 2]
 [5 2 9 1 3 4 7 6 8]
 [4 8 7 6 2 9 5 3 1]
 [2 6 3 4 1 5 9 8 7]
 [9 7 4 8 6 3 1 2 5]
 [8 5 1 7 9 2 6 4 3]
 [1 3 8 9 4 7 2 5 6]
 [6 9 2 3 5 1 8 7 4]
 [7 4 5 2 8 6 3 1 9]]
CPU times: user 68.4 ms, sys: 0 ns, total: 68.4 ms
Wall time: 67.2 ms


### Task 3 : Smart Backtracking Algorithm (40 pts)
Incorporate at least one strategy of minimum remaining values (MRV), least constraining value (LCV), and forward checking in your backtracking algorithm.

<p align="center">
<img src="https://upload.wikimedia.org/wikipedia/commons/8/8c/Sudoku_solved_by_bactracking.gif" width="250" height="250" /></p>


In [90]:
def getGrid(square):
    row,col = square[0], square[1]
    gridsize = 3#int(math.sqrt(len(puzzle)))
    #--- slice the puzzle into grids of gridsize
    slices = [slice(0,gridsize), slice(gridsize,gridsize*2), slice(gridsize*2,gridsize*3)]
    grid = (slices[row//gridsize], slices[col//gridsize])
    #print(grid)
    return grid

"""
Function --- Returns remaining values for a square position of sudoku puzzle
"""
def getRemainingValues(square, puzzle):
    row,col = square[0],square[1]
    #--- values to be used 
    allVals = np.array(range(1,10))
    #--- get all the values of unfilled squares
    #print(puzzle[getGrid(square)])
    values = reduce(np.union1d, (puzzle[row,:], puzzle[:,col], puzzle[getGrid(square)]))
    remaining_values = np.setdiff1d(allVals, values)
    return remaining_values


puzzle = getPuzzle(test2)
print(getRemainingValues([3,3],puzzle))

[[3 0 6 5 0 8 4 0 0]
 [5 2 0 0 0 0 0 0 0]
 [0 8 7 0 0 0 0 3 1]
 [0 0 3 0 1 0 0 8 0]
 [9 0 0 8 6 3 0 0 5]
 [0 5 0 0 9 0 6 0 0]
 [1 3 0 0 0 0 2 5 0]
 [0 0 0 0 0 0 0 7 4]
 [0 0 5 2 0 6 3 0 0]]
[4 7]


In [91]:
"""
# --- MRV - Minimum remaining value
in this approach we go row by row return the one with less number of zeors satisfied with constraint
returns position of minimum remaining value after all the constraints are satisfied. 
"""
def MRV(puzzle):
    remaining_squares = [list(e) for e in np.transpose(np.where(puzzle==0))] #--- get remaining squares where we need to place the number
    remaining_values_list = [getRemainingValues(square, puzzle) for square in remaining_squares] #--- get remaining values for each square where we need to place the number
    remaining_values_lens = [len(remaining_values) for remaining_values in remaining_values_list]#--- get length of remaining values values for each square
    index = np.argmin(remaining_values_lens)                                                     #--- select the one with minimum length i.e. one with minimum reamining values and return its index
    return remaining_squares[index], remaining_values_list[index]                                #--- Return the position of minimum remaining value and list of minimum remaining value for that position  

"""
# --- LCV - least constraining value
counts the number of times a value appears in empty squares
"""
def LCV(puzzle):
    remaining_squares = [list(e) for e in np.transpose(np.where(puzzle==0))] #--- get remaining squares where we need to place the number
    remaining_values_list = [getRemainingValues(square, puzzle) for square in remaining_squares] #--- get remaining values for each square where we need to place the number
    return puzzle

"""
#--- Forward Checking
Implements forward checking algorithm for sudoku takes the puzzle and check if the value placed is a right one or not
"""
def ForwardChecking(puzzle, row, col, value):
    if checkRow(puzzle, row, value):
        if checkCol(puzzle, col, value):
            if checkGrid(puzzle, row, col, value):
                return True                        #--- Return True as all the constrainst are satisfied
    return False                             #--- Return False If any constrainst fails 

def isComplete(puzzle):
    if 0 in puzzle:
        return False
    else:
        return True

def recursive_backtracking(puzzle):
    if isComplete(puzzle):
        return True
    square, values = MRV(puzzle)
    for value in values:
        puzzle[square[0],square[1]] = value
        if recursive_backtracking(puzzle):
            return True
        else:
            puzzle[square[0],square[1]] = 0
    return False

def backtracking_search(puzzle):
    if (recursive_backtracking(puzzle)):
        print('--- Puzzle Solved ---')
        print(puzzle)
    else:
        print('--- Puzzle Unsolved ---')
        print(puzzle)

In [92]:
%%time
backtracking_search(puzzle)

--- Puzzle Solved ---
[[3 1 6 5 7 8 4 9 2]
 [5 2 9 1 3 4 7 6 8]
 [4 8 7 6 2 9 5 3 1]
 [2 6 3 4 1 5 9 8 7]
 [9 7 4 8 6 3 1 2 5]
 [8 5 1 7 9 2 6 4 3]
 [1 3 8 9 4 7 2 5 6]
 [6 9 2 3 5 1 8 7 4]
 [7 4 5 2 8 6 3 1 9]]
CPU times: user 178 ms, sys: 0 ns, total: 178 ms
Wall time: 176 ms


### Forward Checking 

implement a forward checking approach to solve sudoku


<p align="center">
<img src="https://www.csie.ntu.edu.tw/~d91003/ai/project2/forward_checking.gif" width="450" height="250" /></p>


Forward Checking in sudoku


In [93]:
"""
#--- Implementing Forward Checking with MRV
"""
def recursive_backtrackingFC(puzzle):
    if isComplete(puzzle):
        return True
    square, values = MRV(puzzle)
    for value in values:
        puzzle[square[0],square[1]] = value
        if ForwardChecking(puzzle, square[0], square[1], value):
              return True
        else:
              puzzle[square[0],square[1]] = 0
    return False

def backtracking_searchFC(puzzle):
    if (recursive_backtracking(puzzle)):
        print('--- Puzzle Solved ---')
        print(puzzle)
    else:
        print('--- Puzzle Unsolved ---')
        print(puzzle)

In [94]:
puzzle = getPuzzle(test2)

[[3 0 6 5 0 8 4 0 0]
 [5 2 0 0 0 0 0 0 0]
 [0 8 7 0 0 0 0 3 1]
 [0 0 3 0 1 0 0 8 0]
 [9 0 0 8 6 3 0 0 5]
 [0 5 0 0 9 0 6 0 0]
 [1 3 0 0 0 0 2 5 0]
 [0 0 0 0 0 0 0 7 4]
 [0 0 5 2 0 6 3 0 0]]


In [95]:
%%time
backtracking_searchFC(puzzle)

--- Puzzle Solved ---
[[3 1 6 5 7 8 4 9 2]
 [5 2 9 1 3 4 7 6 8]
 [4 8 7 6 2 9 5 3 1]
 [2 6 3 4 1 5 9 8 7]
 [9 7 4 8 6 3 1 2 5]
 [8 5 1 7 9 2 6 4 3]
 [1 3 8 9 4 7 2 5 6]
 [6 9 2 3 5 1 8 7 4]
 [7 4 5 2 8 6 3 1 9]]
CPU times: user 185 ms, sys: 0 ns, total: 185 ms
Wall time: 183 ms


### Task 4 : Analysis and Report

#### For Task 1 : Naive backtracking problem
1. passed sudoku-test-1 and sudoku-test-2 but was not able to solve sudoku-test-final, need to debug initially, but was able to solve it initially.

#### For Task 2 : smart backtracking problem
1. Implemented MRV, minimum remaining values was able to solve every level of the sudoku puzzle in a less time compared to naive back tracking.
2. Implemented Forward Checking with MRV was able to solve every puzzle level but saw a drawback of more time in solving as the puzzle level moves from easy to hard.

below is the table of every approach with its execution time.
<table>
  <tr>
    <th rowspan="2"> Level of game </th> 
    <th colspan="4" align="center"> Performance of Various Backtracking Algorithms </th>
  </tr>
  <tr>
    <th>Naive BackTracking</th>
    <th>Smart BackTracking Using MRV(Minimum Remaining Values)</th>
    <th>Smart BackTracking Using Forward Checking</th>
    <th>Smart BackTracking Using Forward Checking + MRV</th>
  </tr>
  <tr>
    <td>Easy</td>
    <td>232 ms</td>
    <td>133 ms</td>
    <td>246 ms</td>
    <td>127 ms</td>
  </tr>
  <tr>
    <td>Medium</td>
    <td>51.2 ms</td>
    <td>176 ms</td>
    <td>67.2 ms</td>
    <td>183 ms</td>
  </tr>
  <tr>
    <td>Hard</td>
    <td>12.1 ms</td>
    <td>149 ms</td>
    <td>13.8 ms</td>
    <td>124 ms</td>
  </tr>
  <tr>
    <td>Evil</td>
    <td>446 ms</td>
    <td>318 ms</td>
    <td>446 ms</td>
    <td>462 ms</td>
  </tr>
</table>


From the above table we can see that forward checking performs better compared to other algorithms most of the time, but we can also see that as the level of the game raises it is difficult to choose which approach is better.